## <b> Problem Statement </b>
- <b> <i> FoodHunter </i> </b> , a food delivery app, has experienced a drop in revenue over the months. The goal is to investigate the intensity of the decline, identify the causes and contributing factors behind this decline and help to develop strategies to overcome these challenges.

## <b> loading the SQL extension and connecting to the <i> foodhunter database </i> </b>

In [7]:
%load_ext sql
%sql mysql+mysqlconnector://root:root@localhost/

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [2]:
%%sql
use foodhunter;
show tables;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
6 rows affected.


[('customers',),
 ('drivers',),
 ('food_items',),
 ('orders',),
 ('orders_items',),
 ('restaurants',)]

## <b> Investigating tables</b>

### <b> orders</b>

In [30]:
%%sql
#understanding datatype, field characteristics
desc foodhunter.orders;

 * mysql+mysqlconnector://root:***@localhost/
13 rows affected.


[('order_id', 'int', 'YES', '', None, ''),
 ('customer_id', 'bigint', 'YES', '', None, ''),
 ('order_date', 'datetime', 'YES', '', None, ''),
 ('order_time', 'time', 'YES', '', None, ''),
 ('delivered_date', 'text', 'YES', '', None, ''),
 ('delivered_time', 'time', 'YES', '', None, ''),
 ('delivery_address', 'text', 'YES', '', None, ''),
 ('driver_id', 'bigint', 'YES', '', None, ''),
 ('total_price', 'bigint', 'YES', '', None, ''),
 ('delivery_fee', 'double', 'YES', '', None, ''),
 ('discount', 'double', 'YES', '', None, ''),
 ('final_price', 'double', 'YES', '', None, ''),
 ('order_rating', 'double', 'YES', '', None, '')]

In [4]:
%%sql

# identifying the timeframe for our root-cause analysis
select min(order_date) as min_order_date, 
max(order_date) as max_order_date 
from foodhunter.orders;

 * mysql+mysqlconnector://root:***@localhost/
1 rows affected.


[(datetime.datetime(2022, 6, 1, 0, 0), datetime.datetime(2022, 9, 29, 0, 0))]

### observations
* We've a data from 1st Jun 2022 to 29th Sept 2022 to understand why there was a drop in revenue for foodhunter during this timeframe and to identify the underlying causes.

### <b> orders_items</b>

In [5]:
%%sql
desc foodhunter.orders_items;

 * mysql+mysqlconnector://root:***@localhost/
3 rows affected.


[('order_id', 'int', 'YES', '', None, ''),
 ('item_id', 'int', 'YES', '', None, ''),
 ('quantity', 'int', 'YES', '', None, '')]

### <b> restaurants</b>

In [6]:
%%sql
desc foodhunter.restaurants;

 * mysql+mysqlconnector://root:***@localhost/
4 rows affected.


[('restaurant_id', 'int', 'NO', 'PRI', None, ''),
 ('restaurant_name', 'text', 'YES', '', None, ''),
 ('cuisine', 'text', 'YES', '', None, ''),
 ('address', 'text', 'YES', '', None, '')]

### <b> customers</b>

In [7]:
%%sql
desc foodhunter.customers;

 * mysql+mysqlconnector://root:***@localhost/
5 rows affected.


[('customer_id', 'int', 'NO', 'PRI', None, ''),
 ('first_name', 'text', 'YES', '', None, ''),
 ('last_name', 'text', 'YES', '', None, ''),
 ('phone_number', 'text', 'YES', '', None, ''),
 ('address', 'text', 'YES', '', None, '')]

### <b> drivers</b>

In [8]:
%%sql
desc foodhunter.drivers;

 * mysql+mysqlconnector://root:***@localhost/
3 rows affected.


[('driver_id', 'int', 'NO', 'PRI', None, ''),
 ('name', 'text', 'YES', '', None, ''),
 ('rating', 'double', 'YES', '', None, '')]

### <b> food_items</b>

In [9]:
%%sql
desc foodhunter.food_items;

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('item_id', 'int', 'NO', 'PRI', None, ''),
 ('restaurant_id', 'int', 'YES', 'MUL', None, ''),
 ('item_name', 'text', 'YES', '', None, ''),
 ('food_type', 'text', 'YES', '', None, ''),
 ('price', 'text', 'YES', '', None, ''),
 ('calories', 'double', 'YES', '', None, '')]

### <b> The rate of change in order count and revenue over the months </b>

In [20]:
%%sql

with order_revenue_summary as (
  select 
  concat(year(order_date), '-', month(order_date)) as year__month, 
  count(distinct order_id) as ordercount, 
  sum(final_price) as tot_revenue 
  from foodhunter.orders 
  group by concat(year(order_date), '-', month(order_date))
),
lag_order_revenue_summary as(
  select
  year__month,
  ordercount,
  lag(ordercount) over(order by year__month asc) as lag_order_count,
  lead(ordercount, 3) over(order by year__month asc) as lead_order_count,
  tot_revenue,
  lag(tot_revenue) over(order by year__month asc) as lag_tot_revenue,
  lead(tot_revenue, 3) over(order by year__month asc) as lead_tot_revenue
  from order_revenue_summary
)

select 
  year__month,
  ordercount,
  lag_order_count,
  round(((ordercount - lag_order_count) / lag_order_count), 4) as orderCountChangeRate,
  round(((lead_order_count - ordercount) / ordercount), 4) as overallOrderCountChangeRate,
  round(tot_revenue, 2) as tot_revenue,
  round(lag_tot_revenue, 2) as lag_tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue), 4) as totRevenueChangeRate,
  round(((lead_tot_revenue - tot_revenue) / tot_revenue), 4) as overallTotRevenueChangeRate
from lag_order_revenue_summary;

 * mysql+mysqlconnector://root:***@localhost/
4 rows affected.


[('2022-6', 12502, None, None, Decimal('-0.2509'), 347577.5, None, None, -0.2573),
 ('2022-7', 11144, 12502, Decimal('-0.1086'), None, 308601.5, 347577.5, -0.1121, None),
 ('2022-8', 10107, 11144, Decimal('-0.0931'), None, 283365.9, 308601.5, -0.0818, None),
 ('2022-9', 9365, 10107, Decimal('-0.0734'), None, 258161.1, 283365.9, -0.0889, None)]

#### <b> observations </b>
* From the data, we observe a downward trend in order count with <b>25% decline</b> and total revenue with <b>26% decline</b> from June 2022 to September 2022.
  
* <b> The order count </b> shows a 10.86% drop from June to July, 9.31% drop from July to August and 7.34% drop from August to September. While the rate of decline is slowing, the overall drop in order count is still significant.

* On the Otherhand, <b> The total revenue </b> shows an 11.21% drop from June to July, 8.18% drop from July to August and 8.89% drop from August to September. Although the rate of decline fluctuates, the overall decrease in total revenue is significant and denotes requiring immediate attention.

* Based on the rate of change, we can infer that the factors causing the decline are fading out. <i><b>[Assumption] :</b> It seems that recovering phase gets started or marketing strategies have started to adjust and adapt, leading to a reduction in the impact of losses.</i>

* To get the clear picture, It is important to analyze the data at various granularities, as revenue shows signs of volatility <i>[out of control] </i> through its rate of change.

### <b> Weekday vs Weekend: Monthly Variations in Orders and Revenue </b>

In [11]:
%%sql

select
  day_category,
  count(distinct order_id) as ordercount,
  round(sum(final_price), 2) as tot_revenue,
  case when day_category = "Week-Day" 
    then count(distinct order_id) / 5
    else count(distinct order_id) / 2
  end as ordercount_per_dayCat_approx,
  case when day_category = "Week-Day"
    then round(sum(final_price), 2) / 5
    else round(sum(final_price), 2) / 2
  end as tot_revenue_per_dayCat_approx
from
  (select
    *,
    case 
     when dayofweek(order_date) in (1,7) then "Week-End"
     else "Week-Day"
    end as day_category
  from foodhunter.orders) q1
group by day_category;

 * mysql+mysqlconnector://root:***@localhost/
2 rows affected.


[('Week-Day', 31412, 870423.1, Decimal('6282.4000'), 174084.62),
 ('Week-End', 11706, 327282.9, Decimal('5853.0000'), 163641.45)]

#### <b> observations </b>
* Weekdays has more revenue and order count than weekends as expected.

In [21]:
%%sql

with weekpart_order_revenue_summary as(
 select
   concat(year(order_date), '-', month(order_date)) as year__month,
   case
     when dayofweek(order_date) in (1,7) then "Week-End"
       else "Week-Day"
   end as day_category,
   count(distinct order_id) as ordercount, 
   sum(final_price) as tot_revenue 
 from foodhunter.orders
   group by concat(year(order_date), '-', month(order_date)), day_category
),
lag_weekpart_summary as(
  select
    year__month,
    day_category,
    ordercount,
    lag(ordercount) over(partition by day_category order by year__month) as lag_ordercount,
    lead(ordercount, 3) over(partition by day_category order by year__month) as lead_overall_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by day_category order by year__month) as lag_tot_revenue,
    lead(tot_revenue, 3) over(partition by day_category order by year__month) as lead_overall_rev
  from weekpart_order_revenue_summary
)

select 
  year__month,
  day_category,
  round(ordercount, 2) as ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount), 4) as ordercount_change_in_percent,
  round(((lead_overall_ordercount - ordercount)/ ordercount), 4) as ordercount_overall_change_in_percent,
  round(tot_revenue, 2) as tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue), 4) as tot_revenue_change_in_percent,
  round(((lead_overall_rev - tot_revenue)/ tot_revenue), 4) as totRev_overall_change_in_percent
from lag_weekpart_summary;

 * mysql+mysqlconnector://root:***@localhost/
8 rows affected.


[('2022-6', 'Week-Day', 9043, None, Decimal('-0.2382'), 250657.7, None, -0.2452),
 ('2022-7', 'Week-Day', 8096, Decimal('-0.1047'), None, 224993.4, -0.1024, None),
 ('2022-8', 'Week-Day', 7384, Decimal('-0.0879'), None, 205586.2, -0.0863, None),
 ('2022-9', 'Week-Day', 6889, Decimal('-0.0670'), None, 189185.8, -0.0798, None),
 ('2022-6', 'Week-End', 3459, None, Decimal('-0.2842'), 96919.8, None, -0.2883),
 ('2022-7', 'Week-End', 3048, Decimal('-0.1188'), None, 83608.1, -0.1373, None),
 ('2022-8', 'Week-End', 2723, Decimal('-0.1066'), None, 77779.7, -0.0697, None),
 ('2022-9', 'Week-End', 2476, Decimal('-0.0907'), None, 68975.3, -0.1132, None)]

#### <b> observations </b>
* for weekday, ordercounts shows a downward trend with 24% & total revenue shows a downward trend with 25% from June to September, with the rate of decline gradually slowing over the months.
  
* for weekend, ordercounts shows a downward trend with 28% & total revenue shows a downward trend with 29% over the months, with a fluctuations in the rate of decline.

* from this observation, at this level of granularity, we can conclude that weekend revenue across the four months is highly volatile, with significant fluctuations in the rate of change, while both order count and revenue are higher on weekdays than on weekends.

* foodhunter should prioritize on weekend operations as a primary focus than weekday operations due to revenue fluctuations (<i>situation is unpredictable / out of control</i>), as the current strategy does not seem effective for weekends in terms of revenue.

### <b> OrderCount & Total Revenue Variation Across Time Segments </b>

#### <b> <i> Segmenting time periods according to typical eating rhythm </i></b>
- #### <b> <i> order_timings => dining_segment </i></b>
- 07:00 – 10:00	=> Breakfast
- 10:01 – 11:59	=> Brunch
- 12:00 – 14:59	=> Lunch
- 15:00 – 18:59	=> Snacks
- 19:00 – 22:00	=> Dinner
- 22:01 – 06:59	=> After hours

In [12]:
%%sql

create or replace view foodhunter_reports.dining_segment_summary as
select
  year__month,
  day_category,
  dining_segment,
  count(distinct order_id) as ordercount,
  sum(final_price) as tot_revenue,
  sum(discount) as discount
from
  (select
     concat(year(order_date), "-", month(order_date)) as year__month,
     case
        when dayofweek(order_date) in (1,7) then "Week-End"
        else "Week-Day"
     end as day_category,
     case
        when order_time between '07:00:00' and '10:00:00' 
        then '1-Breakfast'
        when order_time between '10:01:00' and '11:59:59' 
        then '2-Brunch'
        when order_time between '12:00:00' and '14:59:59' 
        then '3-Lunch'
        when order_time between '15:00:00' and '18:59:59' 
        then '4-Snacks'
        when order_time between '19:00:00' and '22:00:00' 
        then '5-Dinner'
        when (order_time between '22:01:00' and '23:59:59') 
             or 
             (order_time between 
              '00:00:00' and '06:59:59')  
        then '6-After hours'
     end as dining_segment,
     order_id,
     final_price,
     discount
  from foodhunter.orders) q1
group by year__month, day_category, dining_segment;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.


[]

In [9]:
%%sql
#overall revenue and ordercount shares for dining segments.
select
  dining_segment,
  sum(ordercount) as ordercount,
  round((sum(ordercount) / sum(sum(ordercount)) over(rows between unbounded preceding and unbounded following)), 4) as ordercount_share,
  round(sum(tot_revenue), 2) as tot_revenue,
  round((sum(tot_revenue) / sum(sum(tot_revenue)) over(rows between unbounded preceding and unbounded following)), 4) as total_rev_share
from foodhunter_reports.dining_segment_summary
group by dining_segment
order by tot_revenue desc, ordercount desc;

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('6-After hours', Decimal('13726'), Decimal('0.3183'), 380464.4, 0.3177),
 ('4-Snacks', Decimal('7797'), Decimal('0.1808'), 215378.5, 0.1798),
 ('3-Lunch', Decimal('5950'), Decimal('0.1380'), 167552.6, 0.1399),
 ('1-Breakfast', Decimal('5990'), Decimal('0.1389'), 166993.6, 0.1394),
 ('5-Dinner', Decimal('5890'), Decimal('0.1366'), 162552.1, 0.1357),
 ('2-Brunch', Decimal('3765'), Decimal('0.0873'), 104764.8, 0.0875)]

#### <b> Observations </b>
* The After Hours segment (10 PM to 7 AM) records the highest number of orders and total revenue, which could be attributed by several factors Limited restaurant availability, restricted mobility, Late-night cravings, social events and irregular sleep schedules also contribute to higher demand.
  
* Brunch ranks the lowest in both order count and total revenue possibly due to it's not aligning with the typical eating habits of most people.

* To prevent delays and resource shortages, delivery partners should be strategically allocated based on the proportion of metrics for the dining segments above.

In [11]:
%%sql

with weekdaylag_dining_segment_summary as (
  select
    year__month,
    dining_segment,
    ordercount,
    lag(ordercount) over(partition by dining_segment order by year__month) as lag_ordercount,
    lead(ordercount, 3) over(partition by dining_segment order by year__month) as lead_overall_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by dining_segment order by year__month) as lag_tot_revenue,
    lead(tot_revenue, 3) over(partition by dining_segment order by year__month) as lead_overall_rev,
    discount,
    lag(discount) over(partition by dining_segment order by year__month) as lag_discount,
    lead(discount, 3) over(partition by dining_segment order by year__month) as lead_discount
  from foodhunter_reports.dining_segment_summary 
  where day_category = 'Week-Day')

select
  year__month,
  dining_segment,
  ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount), 4) as ordercount_change_in_percent,
  round(((lead_overall_ordercount - ordercount) / ordercount), 4) as ordercount_overall_change_in_percent,
  round(tot_revenue, 2) as tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue), 4) as tot_revenue_change_in_percent,
  round(((lead_overall_rev - tot_revenue) / tot_revenue), 4) as totRev_overall_change_in_percent,
  round(discount, 2) as discount,
  round(((discount - lag_discount) / lag_discount), 4) as discount_change_in_percent,
  round(((lead_discount - discount) / discount), 4) as discount_overall_change_in_percent
from weekdaylag_dining_segment_summary;

 * mysql+mysqlconnector://root:***@localhost/
24 rows affected.


[('2022-6', '1-Breakfast', 1215, None, Decimal('-0.1786'), 33230.7, None, -0.1802, 4484.8, None, -0.1947),
 ('2022-7', '1-Breakfast', 1144, Decimal('-0.0584'), None, 33229.6, -0.0, None, 4781.4, 0.0661, None),
 ('2022-8', '1-Breakfast', 1018, Decimal('-0.1101'), None, 27481.0, -0.173, None, 3663.0, -0.2339, None),
 ('2022-9', '1-Breakfast', 998, Decimal('-0.0196'), None, 27242.2, -0.0087, None, 3611.8, -0.014, None),
 ('2022-6', '2-Brunch', 806, None, Decimal('-0.2878'), 22517.5, None, -0.3034, 3134.0, None, -0.3364),
 ('2022-7', '2-Brunch', 690, Decimal('-0.1439'), None, 19289.9, -0.1433, None, 2591.6, -0.1731, None),
 ('2022-8', '2-Brunch', 650, Decimal('-0.0580'), None, 17782.7, -0.0781, None, 2365.8, -0.0871, None),
 ('2022-9', '2-Brunch', 574, Decimal('-0.1169'), None, 15686.4, -0.1179, None, 2079.6, -0.121, None),
 ('2022-6', '3-Lunch', 1228, None, Decimal('-0.2077'), 33855.7, None, -0.1984, 4687.8, None, -0.2144),
 ('2022-7', '3-Lunch', 1129, Decimal('-0.0806'), None, 31745.8, -0.0623, None, 4240.2, -0.0955, None),
 ('2022-8', '3-Lunch', 996, Decimal('-0.1178'), None, 28130.3, -0.1139, None, 4046.2, -0.0458, None),
 ('2022-9', '3-Lunch', 973, Decimal('-0.0231'), None, 27138.9, -0.0352, None, 3682.6, -0.0899, None),
 ('2022-6', '4-Snacks', 1656, None, Decimal('-0.2917'), 46596.2, None, -0.3065, 6414.8, None, -0.3324),
 ('2022-7', '4-Snacks', 1500, Decimal('-0.0942'), None, 41610.6, -0.107, None, 5662.4, -0.1173, None),
 ('2022-8', '4-Snacks', 1380, Decimal('-0.0800'), None, 38004.8, -0.0867, None, 5174.2, -0.0862, None),
 ('2022-9', '4-Snacks', 1173, Decimal('-0.1500'), None, 32313.7, -0.1497, None, 4282.8, -0.1723, None),
 ('2022-6', '5-Dinner', 1282, None, Decimal('-0.2785'), 35413.5, None, -0.2738, 4816.0, None, -0.2539),
 ('2022-7', '5-Dinner', 1088, Decimal('-0.1513'), None, 29802.7, -0.1584, None, 4039.8, -0.1612, None),
 ('2022-8', '5-Dinner', 992, Decimal('-0.0882'), None, 28083.4, -0.0577, None, 3905.6, -0.0332, None),
 ('2022-9', '5-Dinner', 925, Decimal('-0.0675'), None, 25717.6, -0.0842, None, 3593.4, -0.0799, None),
 ('2022-6', '6-After hours', 2856, None, Decimal('-0.2136'), 79044.1, None, -0.2272, 10768.4, None, -0.2352),
 ('2022-7', '6-After hours', 2545, Decimal('-0.1089'), None, 69314.8, -0.1231, None, 9170.2, -0.1484, None),
 ('2022-8', '6-After hours', 2348, Decimal('-0.0774'), None, 66104.0, -0.0463, None, 8966.0, -0.0223, None),
 ('2022-9', '6-After hours', 2246, Decimal('-0.0434'), None, 61087.0, -0.0759, None, 8236.0, -0.0814, None)]

In [13]:
%%sql

#revenue and ordercount shares for dining segments for weekday.
with weekday_priority as(
select
a.dining_segment,
a.ordercount,
a.ordercount_share,
q2.overallChange_ordercount,
a.tot_revenue,
a.tot_revenue_share,
q2.overallChange_totRev,
#just setting priority on framing out recovery plan based on severity
(-(1.5*a.tot_revenue_share) * (q2.overallChange_totRev)) as attentionScore, #providing 0.5 units more priority for revenue share from total revenue decline
dense_rank() over(order by tot_revenue desc) as rank_Rev, 
dense_rank() over(order by overallChange_totRev asc) as rank_Overall_Drop_Rev,
dense_rank() over(order by ordercount desc) as rank_orders,
dense_rank() over(order by overallChange_ordercount asc) as rank_Overall_Drop_Orders
from
(select
dining_segment,
sum(ordercount) as ordercount,
(sum(ordercount)/sum(sum(ordercount)) over(rows between unbounded preceding and unbounded following)) as ordercount_share,
sum(tot_revenue) as tot_revenue,
(sum(tot_revenue)/sum(sum(tot_revenue)) over(rows between unbounded preceding and unbounded following)) as tot_revenue_share
from foodhunter_reports.dining_segment_summary
where day_category = "Week-Day"
group by dining_segment) a
left join
(select 
dining_segment, ((lead_ordercount-ordercount)/ordercount) as overallChange_ordercount, ((lead_totRevenue-tot_revenue)/tot_revenue) as overallChange_totRev
from
(select year__month, dining_segment,
ordercount, lead(ordercount,3) over(partition by dining_segment order by year__month) as lead_ordercount,
tot_revenue, lead(tot_revenue,3) over(partition by dining_segment order by year__month) as lead_totRevenue
from foodhunter_reports.dining_segment_summary
where day_category = "Week-Day") q1
) q2
on (a.dining_segment = q2.dining_segment) and (q2.overallChange_ordercount is not null) and (q2.overallChange_totRev is not null))
select
dining_segment, ordercount, ordercount_share, rank_orders, overallChange_ordercount, rank_Overall_Drop_Orders, round(tot_revenue,3) as tot_revenue, round(tot_revenue_share,4) as tot_revenue_share, 
rank_Rev, round(overallChange_totRev, 4) as overallChange_totRev, rank_Overall_Drop_Rev, 
round(attentionScore / (select max(attentionScore) from weekday_priority), 2) as attentionScore
from weekday_priority
order by tot_revenue desc, ordercount desc;

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('6-After hours', Decimal('9995'), Decimal('0.3182'), 1, Decimal('-0.2136'), 4, 275549.9, 0.3166, 1, -0.2272, 4, 1.0),
 ('4-Snacks', Decimal('5709'), Decimal('0.1817'), 2, Decimal('-0.2917'), 1, 158525.3, 0.1821, 2, -0.3065, 1, 0.78),
 ('1-Breakfast', Decimal('4375'), Decimal('0.1393'), 3, Decimal('-0.1786'), 6, 121183.5, 0.1392, 3, -0.1802, 6, 0.35),
 ('3-Lunch', Decimal('4326'), Decimal('0.1377'), 4, Decimal('-0.2077'), 5, 120870.7, 0.1389, 4, -0.1984, 5, 0.38),
 ('5-Dinner', Decimal('4287'), Decimal('0.1365'), 5, Decimal('-0.2785'), 3, 119017.2, 0.1367, 5, -0.2738, 3, 0.52),
 ('2-Brunch', Decimal('2720'), Decimal('0.0866'), 6, Decimal('-0.2878'), 2, 75276.5, 0.0865, 6, -0.3034, 2, 0.36)]

#### <b> Observations on weekdays </b>
* Overall decline gets intensified for the dining segments: Brunch & Snacks with roughly <b> 30% drop </b> in revenue and order count over the entire four months from June 2025 to September 2025.
* The rate of decline in order count for the After Hours & Dinner segment is getting slower over the months, but the revenue shows a considerable increase from August to September. This revenue surge was examined against discount trends to assess interdependency, revealing that no such expected interdependency <i>(<b>Assumption:</b> The revenue decline rate is suddenly accelerating, while the order count decline rate is slowing down due to discount strategies)</i>, the rate of change is accelerating for both revenue and discounts.

* The Snacks segment shows an increase in the rate of decline for both order count and revenue over the months.
* The Breakfast, Brunch and Lunch segments have volatility in the rate of drop in both revenue and order count.
* Arranging Dining Segments based on Overall Decline : <b><i> [Snacks] > [Brunch] > [Dinner] > [After hours] > [Lunch] > [BreakFast] </i> </b>
* Arranging Dining Segments based on Total Revenue : <b><i> [After hours] > [Snacks] > [BreakFast] > [Lunch] > [Dinner] > [Brunch] </i> </b>
---
#### <b> Weightage assessment </b>
* Order of Focus : <b> <i> [After Hours] > [Snacks] > [Dinner] > [Lunch] > [Brunch] > [Breakfast]
* The After Hours segment shows a highest revenue share and comparitively moderate revenue decline, and it needs high attention and effective adjustments to optimize its operations and outcomes. <b>[P0]</b>

* Snack Segment shows higher decline percentage, high revenue next to the 'After Hours' Segment needs next level attention to prevent further loss. <b>[P1]</b>

In [33]:
%%sql

with weekendlag_dining_segment_summary as (
  select 
    year__month,
    dining_segment,
    ordercount,
    lag(ordercount) over(partition by dining_segment order by year__month) as lag_ordercount,
    lead(ordercount, 3) over(partition by dining_segment order by year__month) as lead_overall_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by dining_segment order by year__month) as lag_tot_revenue,
    lead(tot_revenue, 3) over(partition by dining_segment order by year__month) as lead_overall_rev,
    discount,
    lag(discount) over(partition by dining_segment order by year__month) as lag_discount,
    lead(discount, 3) over(partition by dining_segment order by year__month) as lead_discount
  from foodhunter_reports.dining_segment_summary 
  where day_category = 'Week-End')

select
  year__month,
  dining_segment,
  ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount), 4) as ordercount_change_in_percent,
  round(((lead_overall_ordercount - ordercount) / ordercount), 4) as ordercount_overall_change_in_percent,
  round(tot_revenue, 2) as tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue), 4) as tot_revenue_change_in_percent,
  round(((lead_overall_rev - tot_revenue) / tot_revenue), 4) as totRev_overall_change_in_percent,
  round(discount, 2) as discount,
  round(((discount - lag_discount) / lag_discount), 4) as discount_change_in_percent,
  round(((lead_discount - discount) / discount), 4) as discount_overall_change_in_percent
from weekendlag_dining_segment_summary;

 * mysql+mysqlconnector://root:***@localhost/
24 rows affected.


[('2022-6', '1-Breakfast', 485, None, Decimal('-0.2680'), 13627.1, None, -0.2173, 1972.4, None, -0.2131),
 ('2022-7', '1-Breakfast', 423, Decimal('-0.1278'), None, 11492.2, -0.1567, None, 1510.8, -0.234, None),
 ('2022-8', '1-Breakfast', 352, Decimal('-0.1678'), None, 10024.3, -0.1277, None, 1372.2, -0.0917, None),
 ('2022-9', '1-Breakfast', 355, Decimal('0.0085'), None, 10666.5, 0.0641, None, 1552.0, 0.131, None),
 ('2022-6', '2-Brunch', 322, None, Decimal('-0.3385'), 9342.3, None, -0.3694, 1352.2, None, -0.4032),
 ('2022-7', '2-Brunch', 264, Decimal('-0.1801'), None, 7157.6, -0.2339, None, 971.4, -0.2816, None),
 ('2022-8', '2-Brunch', 246, Decimal('-0.0682'), None, 7096.9, -0.0085, None, 1037.6, 0.0681, None),
 ('2022-9', '2-Brunch', 213, Decimal('-0.1341'), None, 5891.5, -0.1698, None, 807.0, -0.2222, None),
 ('2022-6', '3-Lunch', 472, None, Decimal('-0.3220'), 13442.3, None, -0.3467, 1900.2, None, -0.3849),
 ('2022-7', '3-Lunch', 436, Decimal('-0.0763'), None, 12517.6, -0.0688, None, 1676.4, -0.1178, None),
 ('2022-8', '3-Lunch', 396, Decimal('-0.0917'), None, 11939.8, -0.0462, None, 1727.2, 0.0303, None),
 ('2022-9', '3-Lunch', 320, Decimal('-0.1919'), None, 8782.2, -0.2645, None, 1168.8, -0.3233, None),
 ('2022-6', '4-Snacks', 603, None, Decimal('-0.3068'), 16235.5, None, -0.3072, 2173.0, None, -0.2781),
 ('2022-7', '4-Snacks', 559, Decimal('-0.0730'), None, 15048.0, -0.0731, None, 1987.0, -0.0856, None),
 ('2022-8', '4-Snacks', 508, Decimal('-0.0912'), None, 14322.3, -0.0482, None, 2005.2, 0.0092, None),
 ('2022-9', '4-Snacks', 418, Decimal('-0.1772'), None, 11247.4, -0.2147, None, 1568.6, -0.2177, None),
 ('2022-6', '5-Dinner', 467, None, Decimal('-0.2420'), 12602.6, None, -0.1997, 1752.4, None, -0.1755),
 ('2022-7', '5-Dinner', 409, Decimal('-0.1242'), None, 11082.2, -0.1206, None, 1516.8, -0.1344, None),
 ('2022-8', '5-Dinner', 373, Decimal('-0.0880'), None, 9763.9, -0.119, None, 1397.6, -0.0786, None),
 ('2022-9', '5-Dinner', 354, Decimal('-0.0509'), None, 10086.2, 0.033, None, 1444.8, 0.0338, None),
 ('2022-6', '6-After hours', 1110, None, Decimal('-0.2649'), 31670.0, None, -0.2958, 4339.0, None, -0.3121),
 ('2022-7', '6-After hours', 957, Decimal('-0.1378'), None, 26310.5, -0.1692, None, 3582.0, -0.1745, None),
 ('2022-8', '6-After hours', 848, Decimal('-0.1139'), None, 24632.5, -0.0638, None, 3482.0, -0.0279, None),
 ('2022-9', '6-After hours', 816, Decimal('-0.0377'), None, 22301.5, -0.0946, None, 2985.0, -0.1427, None)]

In [52]:
%%sql

#revenue and ordercount shares for dining segments for weekday.
with weekend_priority as(
select
a.dining_segment,
a.ordercount,
a.ordercount_share,
q2.overallChange_ordercount,
a.tot_revenue,
a.tot_revenue_share,
q2.overallChange_totRev,
#just to set priority on framing out recovery plan based on severity
(-(1.5*a.tot_revenue_share) * (q2.overallChange_totRev)) as attentionScore, #providing 0.5 units more priority for revenue share from total revenue decline
dense_rank() over(order by tot_revenue desc) as rank_Rev, 
dense_rank() over(order by overallChange_totRev asc) as rank_Overall_Drop_Rev,
dense_rank() over(order by ordercount desc) as rank_orders,
dense_rank() over(order by overallChange_ordercount asc) as rank_Overall_Drop_Orders
from
(select
dining_segment,
sum(ordercount) as ordercount,
(sum(ordercount)/sum(sum(ordercount)) over(rows between unbounded preceding and unbounded following)) as ordercount_share,
sum(tot_revenue) as tot_revenue,
(sum(tot_revenue)/sum(sum(tot_revenue)) over(rows between unbounded preceding and unbounded following)) as tot_revenue_share
from foodhunter_reports.dining_segment_summary
where day_category = "Week-End"
group by dining_segment) a
left join
(select dining_segment, ((lead_ordercount-ordercount)/ordercount) as overallChange_ordercount, ((lead_totRevenue-tot_revenue)/tot_revenue) as overallChange_totRev
from
(select year__month, dining_segment,
ordercount, lead(ordercount,3) over(partition by dining_segment order by year__month) as lead_ordercount,
tot_revenue, lead(tot_revenue,3) over(partition by dining_segment order by year__month) as lead_totRevenue
from foodhunter_reports.dining_segment_summary
where day_category = "Week-End") q1
) q2
on a.dining_segment = q2.dining_segment and q2.overallChange_ordercount is not null and q2.overallChange_totRev is not null)
select
dining_segment, ordercount, ordercount_share, rank_orders, overallChange_ordercount, rank_Overall_Drop_Orders, round(tot_revenue,3) as tot_revenue, round(tot_revenue_share, 4) as tot_revenue_share, 
rank_Rev, round(overallChange_totRev,4) as overallChange_totRev, rank_Overall_Drop_Rev, 
round(attentionScore / (select max(attentionScore) from weekend_priority), 2) as attentionScore
from weekend_priority
order by tot_revenue desc, ordercount desc;

 * mysql+mysqlconnector://root:***@localhost/
6 rows affected.


[('6-After hours', Decimal('3731'), Decimal('0.3187'), 1, Decimal('-0.2649'), 5, 104914.5, 0.3206, 1, -0.2958, 4, 1.0),
 ('4-Snacks', Decimal('2088'), Decimal('0.1784'), 2, Decimal('-0.3068'), 3, 56853.2, 0.1737, 2, -0.3072, 3, 0.56),
 ('3-Lunch', Decimal('1624'), Decimal('0.1387'), 3, Decimal('-0.3220'), 2, 46681.9, 0.1426, 3, -0.3467, 2, 0.52),
 ('1-Breakfast', Decimal('1615'), Decimal('0.1380'), 4, Decimal('-0.2680'), 4, 45810.1, 0.14, 4, -0.2173, 5, 0.32),
 ('5-Dinner', Decimal('1603'), Decimal('0.1369'), 5, Decimal('-0.2420'), 6, 43534.9, 0.133, 5, -0.1997, 6, 0.28),
 ('2-Brunch', Decimal('1045'), Decimal('0.0893'), 6, Decimal('-0.3385'), 1, 29488.3, 0.0901, 6, -0.3694, 1, 0.35)]

#### <b> Observations on weekends</b>
* Overall decline gets intensified for the dining segments: Brunch & Lunch with roughly <b> 35% drop </b> in revenue and ordercount over the entire four months from June 2022 to September 2022.
* During weekends, dining segment - brunch experience fluctuations in the rate of decline in revenue and order count over the four-month period.
* Breakfast shows positive signs from August 2022 to September 2022, as the rate of decline slows down and begins to shift towards rate of elevation in terms of both revenue and order count over the period of time.
* The dinner segment shows positive signs over the months, as the rate of decline in revenue and order count becomes slower over time.
* Lunch and snacks segments show a sharp decline over the months in terms of both order counts and revenue, getting faster as time goes on. It has a higher revenue and order share, making it an important segment to consider for framing recovery strategies.
* The decline in order count for the after-hours segment is slowing, but revenue is falling faster from Aug 2022 to Sep 2022. <i> We investigated the case does marketing strategies like offering discounts causing this? but there's no expected relationships between its trends. </i>

<b> Expected <i> if discount offerings causing this </i> : </b> If the rate of decline is getting slower for ordercount and getting faster for revenue will results in slowering in discount offerings

* Ordering Dining Segments based on Overall Decline : <b><i>[Brunch] > [Lunch] > [Snacks] > [After Hours] > [Breakfast] > [Dinner] </i></b>
* Ordering Dining Segments based on Total Revenue : <b><i> [After Hours] > [Snacks] > [Lunch] > [Breakfast] > [Dinner] > [Brunch]</i></b>
---
#### <b> Weightage assessment </b>
* Order of Focus : <b> <i> [After Hours] > [Snacks] > [Lunch] > [Brunch] > [Breakfast] > [Dinner]
  
* After Hours segment shows highest revenue share and also higher decline percentage next to the segments Snacks & Lunch.<b><i> Recommending to treat this as [P0] risk </i></b>
* Snacks and Lunch Segments have higher revenue share and higher decline percentage over the months Jun to Sep. <b> [P1] </b>

### <b> Impact of Discounts on Order Count and Total Revenue </b>

In [15]:
%%sql
select
  month(order_date) as month_no,
  round(avg(discount), 2) as avg_discount,
  round((sum(discount) / sum(final_price))* 100, 3) as discount_ratio_,
  count(distinct order_id) as ordercount,
  round(sum(final_price), 2) as tot_revenue
from foodhunter.orders
group by month(order_date);

 * mysql+mysqlconnector://root:***@localhost/
4 rows affected.


[(6, 3.82, 13.751, 12502, 347577.5),
 (7, 3.74, 13.522, 11144, 308601.5),
 (8, 3.87, 13.813, 10107, 283365.9),
 (9, 3.74, 13.562, 9365, 258161.1)]

#### <b> observation</b>
* By looking at high level, we can safely say that offering discounts does not influence total revenue & the number of orders.
  
* To confirm this, we should perform causality test between 'avg_discount' and 'ordercount' at week level over the entire timeframe. 

### <b> Avg delivery time & Avg order rating & driver rating</b>

In [33]:
%%sql

create or replace view foodhunter_reports.deliverytime_and_revenue as
  select
    order_id,
    order_date,
    case when dayofweek(order_date) in (1,7) then "Week-End" else "Week-Day" end as day_category,
    customer_id,
    o.driver_id,
    order_time,
    delivered_time,
    time_to_sec(timediff(delivered_time, order_time))/60 as delivery_duration_in_mins,
    final_price,
    delivery_fee,
    order_rating,
    rating as driver_rating
  from foodhunter.orders o left join foodhunter.drivers d on o.driver_id = d.driver_id;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.


[]

In [34]:
%%sql

with month_summary as (
select
q1.month_no,
q1.day_category,
q1.avg_delivery_duration,
q2.median_delivery_duration,
q1.avg_order_rating,
q1.ordercount,
q1.total_revenue
from
(select
  month(order_date) as month_no,
  day_category,
  avg(delivery_duration_in_mins) as avg_delivery_duration,
  avg(order_rating) as avg_order_rating,
  count(distinct order_id) as ordercount,
  sum(final_price) as total_revenue
 from foodhunter_reports.deliverytime_and_revenue
 group by month(order_date), day_category
) q1 inner join
(
with ranking as(
select order_date, day_category, delivery_duration_in_mins, rnk, count(*) over(partition by month(order_date), day_category) as total_rows, max(rnk) over(partition by month(order_date), day_category) as max_rnk
from
(select order_date, day_category, delivery_duration_in_mins, row_number() over (partition by month(order_date), day_category order by delivery_duration_in_mins) as rnk from foodhunter_reports.deliverytime_and_revenue) q
)
select
month(order_date) as month_no,
day_category,
case when total_rows % 2 = 1 then max(case when rnk = (total_rows + 1)/2 then delivery_duration_in_mins end) else avg(case when rnk in(total_rows/2, (total_rows/2)+1) then delivery_duration_in_mins end) end as median_delivery_duration
from ranking
group by month(order_date), day_category
) q2 on q1.month_no = q2.month_no and q1.day_category = q2.day_category
),
month_lagsummary as(
select
month_no,
day_category,
avg_delivery_duration,
lag(avg_delivery_duration) over(partition by day_category order by month_no) as lag_avg_delivery_duration,
median_delivery_duration,
lag(median_delivery_duration) over(partition by day_category order by month_no) as lag_median_delivery_duration,
avg_order_rating,
lag(avg_order_rating) over(partition by day_category order by month_no) as lag_order_rating,
ordercount,
lag(ordercount) over(partition by day_category order by month_no) as lag_ordercount,
total_revenue,
lag(total_revenue) over(partition by day_category order by month_no) as lag_revenue
from
month_summary
)
select
month_no,
day_category,
round(avg_delivery_duration,3) as avg_delivery_duration,
round(((avg_delivery_duration - lag_avg_delivery_duration) / (lag_avg_delivery_duration))*100, 2) as delivery_duration_change_percent_avg,
round(median_delivery_duration,3) as median_delivery_duration,
round(((median_delivery_duration - lag_median_delivery_duration) / (lag_median_delivery_duration))*100, 2) as delivery_duration_change_percent_med,
round(avg_order_rating,2) as avg_order_rating,
round(((avg_order_rating - lag_order_rating) / (lag_order_rating))*100, 2) as order_rating_change_percent,
ordercount,
round(((ordercount - lag_ordercount) / (lag_ordercount))*100, 2) as ordercount_change_percent,
round(total_revenue, 2) as total_revenue,
round(((total_revenue - lag_revenue) / lag_revenue)*100, 2) as total_revenue_change_percent
from
month_lagsummary
order by day_category, month_no;

 * mysql+mysqlconnector://root:***@localhost/
8 rows affected.


[(6, 'Week-Day', Decimal('19.130'), None, Decimal('19.000'), None, 4.22, None, 9043, None, 250657.7, None),
 (7, 'Week-Day', Decimal('21.130'), Decimal('10.45'), Decimal('21.000'), Decimal('10.53'), 3.71, -12.02, 8096, Decimal('-10.47'), 224993.4, -10.24),
 (8, 'Week-Day', Decimal('24.163'), Decimal('14.36'), Decimal('24.000'), Decimal('14.29'), 3.38, -8.95, 7384, Decimal('-8.79'), 205586.2, -8.63),
 (9, 'Week-Day', Decimal('31.173'), Decimal('29.01'), Decimal('31.000'), Decimal('29.17'), 2.99, -11.44, 6889, Decimal('-6.70'), 189185.8, -7.98),
 (6, 'Week-End', Decimal('19.206'), None, Decimal('19.000'), None, 4.22, None, 3459, None, 96919.8, None),
 (7, 'Week-End', Decimal('21.176'), Decimal('10.25'), Decimal('21.000'), Decimal('10.53'), 3.74, -11.39, 3048, Decimal('-11.88'), 83608.1, -13.73),
 (8, 'Week-End', Decimal('24.040'), Decimal('13.53'), Decimal('24.000'), Decimal('14.29'), 3.38, -9.55, 2723, Decimal('-10.66'), 77779.7, -6.97),
 (9, 'Week-End', Decimal('30.968'), Decimal('28.82'), Decimal('31.000'), Decimal('29.17'), 3.01, -10.78, 2476, Decimal('-9.07'), 68975.3, -11.32)]

#### <b> observation</b>
* We observe inverse relationship between delivery duration and order count, tot_revenue as well as order rating. This suggests that delivery duration acts as one of the factors which impact revenue, order count and order rating.
  
* To confirm this, we should perform causality test between delivery_duration and Order Count/ Total Revenue/ Order Rating.

### <b> Food Items preferences</b>

In [35]:
%%sql

select
fi.food_type, count(distinct o.order_id) as ordercount, count(distinct customer_id) as customer_count, sum(oi.quantity) as quantity_sum, count(distinct fi.item_id) as item_count, count(distinct r.cuisine) as cuisine_count
from foodhunter.orders o
left join foodhunter.orders_items oi using(order_id)
left join foodhunter.food_items fi using(item_id)
left join foodhunter.restaurants r using(restaurant_id)
group by fi.food_type;

 * mysql+mysqlconnector://root:***@localhost/
2 rows affected.


[('Non-Veg', 34664, 9723, Decimal('63077'), 219, 4),
 ('Veg', 15454, 7903, Decimal('23239'), 81, 4)]

In [36]:
%%sql

update foodhunter.food_items 
set food_type = case when upper(food_type) like 'v%' then 'Veg'
when upper(food_type) like 'non-%' then 'Non-Veg' else 'others'
end;

 * mysql+mysqlconnector://root:***@localhost/
314 rows affected.


[]

In [37]:
%%sql

select
fi.food_type, count(distinct o.order_id) as ordercount, count(distinct customer_id) as customer_count, sum(oi.quantity) as quantity_sum, count(distinct fi.item_id) as item_count, count(distinct r.cuisine) as cuisine_count
from foodhunter.orders o
left join foodhunter.orders_items oi using(order_id)
left join foodhunter.food_items fi using(item_id)
left join foodhunter.restaurants r using(restaurant_id)
group by fi.food_type;

 * mysql+mysqlconnector://root:***@localhost/
2 rows affected.


[('Non-Veg', 34664, 9723, Decimal('63077'), 219, 4),
 ('Veg', 15454, 7903, Decimal('23239'), 81, 4)]

In [48]:
%%sql
#Items that have not been ordered by the customers over the 4 months from June to September
#restaurants with zero orders through foodhunter
select
item_id, item_name, food_type, restaurant_id, restaurant_name, cuisine, price
from foodhunter.orders_items 
right join foodhunter.food_items using(item_id)
right join foodhunter.restaurants using(restaurant_id)
where order_id is null;

 * mysql+mysqlconnector://root:***@localhost/
14 rows affected.


[(401, 'Pad Sator', 'Non-Veg', 4, 'Minetta Tavern', 'Thai', None),
 (402, 'Kana Moo Krob', 'Non-Veg', 4, 'Minetta Tavern', 'Thai', None),
 (403, 'Nam Tok', 'Non-Veg', 4, 'Minetta Tavern', 'Thai', None),
 (406, 'Special Pasta', 'Non-Veg', 13, "Roberta's Pizza", 'Italian', None),
 (407, 'Pepperoni pizza', 'Non-Veg', 13, "Roberta's Pizza", 'Italian', None),
 (408, 'Papadum', 'Veg', 16, 'La Vara', 'Italian', None),
 (409, 'Fiano di Avellino', 'Veg', 35, 'Buvette', 'Italian', None),
 (410, 'Cozze al vino bianco', 'Non-Veg', 35, 'Buvette', 'Italian', None),
 (411, 'Special Pasta', 'Non-Veg', 58, 'Le Coucou', 'Italian', None),
 (412, 'Bellini', 'Veg', 58, 'Le Coucou', 'Italian', None),
 (413, 'Special Pasta', 'Non-Veg', 60, 'Frenchette', 'Italian', None),
 (404, 'Kana Moo Krob', 'Non-Veg', 75, 'Pam Real Thai', 'Thai', None),
 (405, 'Tod Man Pla', 'Non-Veg', 75, 'Pam Real Thai', 'Thai', None),
 (414, 'Pepperoni pizza', 'Non-Veg', 99, 'Egg Shop', 'Indian', None)]

### <b> customer demographics analysis</b>

In [50]:
%%sql
# Checking for SCD (conflicts in customer address between old and new records)
select count(*)
from foodhunter.orders o
right join foodhunter.customers c on o.customer_id = c.customer_id
where o.delivery_address <> c.address;

 * mysql+mysqlconnector://root:***@localhost/
1 rows affected.


[(0,)]

In [107]:
%%sql
#ordercount & revenue based on customer_location
select
address_ as customer_location,
count(distinct order_id) as total_orders,
round(sum(final_price),2) as total_revenue
from
(select customer_id, lower(trim(regexp_replace(address, '[0-9]', ''))) as address_ from foodhunter.customers) c
join orders o using(customer_id)
group by c.address_
order by total_revenue desc, total_orders desc
limit 30;

 * mysql+mysqlconnector://root:***@localhost/
30 rows affected.


[('parkview avenue', 343, 10890.4),
 ('oak street', 364, 10807.2),
 ('greenfield terrace', 368, 10781.1),
 ('washington avenue', 354, 10674.9),
 ('chestnut avenue', 376, 10491.0),
 ('beechwood road', 357, 10430.9),
 ('high street', 345, 10206.3),
 ('glenwood avenue', 333, 10194.9),
 ('northview road', 376, 10152.0),
 ('brookdale place', 339, 10108.4),
 ('rosedale avenue', 337, 10085.1),
 ('cedar lane', 323, 10031.2),
 ('fairway drive', 335, 10007.1),
 ('hickory lane', 365, 9994.8),
 ('cedar street', 352, 9968.4),
 ('linden lane', 348, 9954.3),
 ('jefferson avenue', 326, 9941.2),
 ('locust street', 367, 9897.0),
 ('vine street', 363, 9874.5),
 ('lakeview drive', 343, 9664.6),
 ('springdale road', 333, 9638.0),
 ('elmwood terrace', 334, 9616.3),
 ('mainview avenue', 338, 9603.2),
 ('sunnyside road', 359, 9596.9),
 ('lincoln circle', 327, 9595.1),
 ('broad street', 349, 9571.6),
 ('park lane', 334, 9570.0),
 ('hillcrest drive', 336, 9569.0),
 ('riverside drive', 326, 9565.5),
 ('cedar road', 342, 9548.7)]

In [108]:
%%sql
#ordercount and total_revenue changes over the months
select
month(o.order_date) as month,
address_ as customer_location,
count(distinct order_id) as total_orders,
round(sum(final_price),2) as total_revenue
from
(select customer_id, lower(trim(regexp_replace(address, '[0-9]', ''))) as address_ from foodhunter.customers) c
join orders o using(customer_id)
where lower(c.address_) in ("parkview avenue", "oak street", "greenfield terrace", "washington avenue", "glenwood avenue")
group by month(o.order_date), c.address_
order by customer_location asc, month asc;

 * mysql+mysqlconnector://root:***@localhost/
20 rows affected.


[(6, 'glenwood avenue', 117, 3823.3),
 (7, 'glenwood avenue', 85, 2628.7),
 (8, 'glenwood avenue', 63, 1942.3),
 (9, 'glenwood avenue', 68, 1800.6),
 (6, 'greenfield terrace', 98, 3008.4),
 (7, 'greenfield terrace', 107, 3050.4),
 (8, 'greenfield terrace', 86, 2334.5),
 (9, 'greenfield terrace', 77, 2387.8),
 (6, 'oak street', 84, 2348.1),
 (7, 'oak street', 93, 3167.1),
 (8, 'oak street', 86, 2573.4),
 (9, 'oak street', 101, 2718.6),
 (6, 'parkview avenue', 99, 3367.0),
 (7, 'parkview avenue', 79, 2228.1),
 (8, 'parkview avenue', 85, 2610.1),
 (9, 'parkview avenue', 80, 2685.2),
 (6, 'washington avenue', 119, 3121.0),
 (7, 'washington avenue', 76, 2425.3),
 (8, 'washington avenue', 67, 2182.7),
 (9, 'washington avenue', 92, 2945.9)]

### <b> Meal Pricing Analysis </b>

#### <b> <i> Segmenting orders according to the total price of the food items per meal</i></b>
- #### <b> <i> total_price range => Class </i></b>
- `1$ - 10$` => Class A
- `11$ - 25$` => Class B
- `26$ - 60$` => Class C
- `61$ - 100$` => Class D
- `> 101$`  => Class E

In [14]:
%%sql

create or replace view foodhunter_reports.price_category_ordercount_revenue as
select
price_category,
round(total_price,2) as total_price,
round(total_price / sum(total_price) over(rows between unbounded preceding and unbounded following), 4) as total_price_share,
round(discount_amt,2) as discount_amt,
round(delivery_fee,2) as delivery_fee,
round(total_revenue,2) as total_revenue,
round(total_revenue / sum(total_revenue) over(rows between unbounded preceding and unbounded following), 4) as total_revenue_share,
dense_rank() over(order by total_revenue desc) as total_revenue_rank,
ordercount,
round(ordercount / sum(ordercount) over(rows between unbounded preceding and unbounded following), 4) as ordercount_share,
dense_rank() over(order by ordercount desc) as ordercount_rank,
round(total_revenue / ordercount, 2) as price_per_order
from
  (select
    price_category,
    count(distinct order_id) as ordercount,
    sum(final_price) as total_revenue,
    sum(discount) as discount_amt,
    sum(total_price) as total_price,
    sum(delivery_fee) as delivery_fee
   from
     (select
       order_id, final_price, discount, total_price, delivery_fee,
       case when total_price between 1 and 10 then "Class A"
         when total_price between 11 and 25 then "Class B"
         when total_price between 26 and 60 then "Class C"
         when total_price between 61 and 100 then "Class D"
         when total_price >= 101 then "Class E"
       end as price_category
      from foodhunter.orders) q1
   group by price_category) q2
order by price_category;

select
*
from
foodhunter_reports.price_category_ordercount_revenue;

 * mysql+mysqlconnector://root:***@localhost/
0 rows affected.
5 rows affected.


[('Class A', Decimal('77452'), Decimal('0.0630'), 4272.2, 54565.0, 127744.8, 0.1067, 5, 10913, Decimal('0.2531'), 3, 11.71),
 ('Class B', Decimal('286959'), Decimal('0.2334'), 29692.2, 71246.5, 328513.3, 0.2743, 2, 16538, Decimal('0.3836'), 1, 19.86),
 ('Class C', Decimal('443281'), Decimal('0.3605'), 61531.4, 6047.5, 387797.1, 0.3238, 1, 11817, Decimal('0.2741'), 2, 32.82),
 ('Class D', Decimal('190145'), Decimal('0.1546'), 21846.2, 0.0, 168298.8, 0.1405, 4, 2162, Decimal('0.0501'), 4, 77.84),
 ('Class E', Decimal('231690'), Decimal('0.1884'), 46338.0, 0.0, 185352.0, 0.1548, 3, 1688, Decimal('0.0391'), 5, 109.81)]

#### <b> observations</b>
* Most of our orders are from Class B, while most of our revenue comes from Class C.

* Food items & cuisines that belong to Class B and Class C play a major role in our business, followed by Class A and Class E.

In [92]:
%%sql
with orderClassification as
(select
month(order_date) as month_no, price_category,
count(distinct order_id) as order_count, sum(final_price) as tot_revenue 
from
(select
order_date, order_id, final_price,
case when total_price between 1 and 10 then "Class A"
when total_price between 11 and 25 then "Class B"
when total_price between 26 and 60 then "Class C"
when total_price between 61 and 100 then "Class D"
when total_price >= 101 then "Class E"
end as price_category
from foodhunter.orders) q1
group by month_no, price_category),
lagOperations as
(select
month_no, price_category,
order_count, lag(order_count) over(partition by price_category order by month_no) as lag_order_count, tot_revenue, lag(tot_revenue) over(partition by price_category order by month_no) as lag_tot_revenue
from orderClassification
)
select
month_no,
price_category,
order_count,
round(((order_count - lag_order_count) / lag_order_count)*100, 2) as order_count_decline_rate,
tot_revenue,
round(((tot_revenue - lag_tot_revenue)/lag_tot_revenue)*100,2) as tot_rev_decline_rate
from
lagOperations
order by price_category asc, month_no asc;

 * mysql+mysqlconnector://root:***@localhost/
20 rows affected.


[(6, 'Class A', 3113, None, 36540.40000000032, None),
 (7, 'Class A', 2821, Decimal('-9.38'), 32977.40000000041, -9.75),
 (8, 'Class A', 2597, Decimal('-7.94'), 30157.200000000365, -8.55),
 (9, 'Class A', 2382, Decimal('-8.28'), 28069.80000000028, -6.92),
 (6, 'Class B', 4863, None, 96530.70000000006, None),
 (7, 'Class B', 4267, Decimal('-12.26'), 84762.70000000019, -12.19),
 (8, 'Class B', 3831, Decimal('-10.22'), 76080.59999999979, -10.24),
 (9, 'Class B', 3577, Decimal('-6.63'), 71139.29999999993, -6.49),
 (6, 'Class C', 3391, None, 110530.6000000004, None),
 (7, 'Class C', 3074, Decimal('-9.35'), 101658.80000000045, -8.03),
 (8, 'Class C', 2742, Decimal('-10.80'), 89757.30000000013, -11.71),
 (9, 'Class C', 2610, Decimal('-4.81'), 85850.40000000029, -4.35),
 (6, 'Class D', 644, None, 49812.60000000008, None),
 (7, 'Class D', 576, Decimal('-10.56'), 44704.20000000008, -10.26),
 (8, 'Class D', 505, Decimal('-12.33'), 39778.80000000001, -11.02),
 (9, 'Class D', 437, Decimal('-13.47'), 34003.20000000002, -14.52),
 (6, 'Class E', 491, None, 54163.19999999994, None),
 (7, 'Class E', 406, Decimal('-17.31'), 44498.399999999965, -17.84),
 (8, 'Class E', 432, Decimal('6.40'), 47591.99999999997, 6.95),
 (9, 'Class E', 359, Decimal('-16.90'), 39098.399999999994, -17.85)]